### PR TITLE
Fix some potential buffer overflows

### PR DESCRIPTION
**Pull request by kgaillot**

### CODE REVIEW

**Comment by kgaillot** [comment]
@nrwahl2 , can you review this when you get a chance? Thanks

**Comment by kgaillot** [comment]
FYI @AdamKorcz , these are the fixes for the Pacemaker buffer overflows

File: lib/common/iso8601.c [hunk 1]
```diff
@@ -1027,53 +1027,58 @@ parse_date(const char *date_str)
...
         if (fraction) {
-            intermediate = (str[offset] - '0') / (10 ^ lpc);
+            /* @TODO Previously, the code here essentially used:
+             *
+             *     *result += digit / (10 ^ lpc);
+             *
+             * With integer division, lpc > 0 will always yield a zero
+             * expression, and it should be 10 ^ (lpc + 1) anyway. Rather than
+             * fix it, we really just need to replace our ISO 8601 code with
+             * GDateTime.
```
Comment (nrwahl2) [inline lib/common/iso8601.c:1077]: Not to mention that's a bitwise XOR rather than exponentiation

Comment (kgaillot) [reply lib/common/iso8601.c:1077]: lol, I completely overlooked that

**Comment by AdamKorcz** [comment]
@davidkorczynski cc

**Comment by DavidKorczynski** [comment]
I can confirm this fixes the overflows (found by fuzzers in https://github.com/ClusterLabs/pacemaker/pull/3466)

File: lib/common/iso8601.c (hunk 1 above)
Comment (kgaillot) [reply lib/common/iso8601.c]: updated comment

**Comment by kgaillot** [comment]
Updated unit tests for differences in *BSD strftime() implementation

File: lib/common/tests/iso8601/pcmk__time_format_hr_test.c [hunk 2]
```diff
@@ -73,6 +73,11 @@ no_specifiers(void **state)
                      "this has a literal %01 in it", NULL);
     assert_hr_format("%2 this starts and ends with %",
                      "%2 this starts and ends with %", NULL);
+
+    /* strftime() treats % with a number (and no specifier) as a literal string
+     * to be formatted with a field width (undocumented and probably a bug ...)
+     */
+    assert_hr_format("this ends with %10", "this ends with        %10", NULL);
```
Comment (nrwahl2) [inline lib/common/tests/iso8601/pcmk__time_format_hr_test.c:80]: ...what

agreed probably a bug

File: lib/common/tests/iso8601/pcmk__time_format_hr_test.c [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    assert_hr_format(NULL, NULL, NULL); // for pcmk__time_format_hr(NULL, hr)
+}
+
+static void
+no_specifiers(void **state)
+{
+    assert_hr_format("no specifiers", "no specifiers", NULL);
+    assert_hr_format("this has a literal % in it",
+                     "this has a literal % in it",
+                     // *BSD strftime() strips single %
```
Comment (nrwahl2) [inline lib/common/tests/iso8601/pcmk__time_format_hr_test.c:70]: Apparently it also strips `%0`
```
14:24:32  FAIL: pcmk__time_format_hr_test 2 - no_specifiers
14:24:32  # "this has a literal 1 in it" != "this has a literal %01 in it"
```

File: lib/common/tests/iso8601/pcmk__time_format_hr_test.c [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    // strftime() treats %% as a literal %
+    assert_hr_format("%%H %%N", "%H %N");
+}
+
+static void
+with_nano(void **state)
+{
+    assert_hr_format("%H:%M:%S.%06N", "03:04:05.000000");
+    assert_hr_format("The time is %H:%M:%S.%06N right NOW",
+                     "The time is 03:04:05.000000 right NOW");
```
[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/common/tests/iso8601/pcmk__time_format_hr_test.c]: Unfortunately we have no exposed API for that other than the "now" function, which obviously won't work in a unit test. But showing that the format is correct is the main thing here.

File: lib/common/iso8601.c [hunk 5]
```diff
@@ -1916,36 +1916,53 @@ pcmk__time_hr_free(pcmk__time_hr_t * hr_dt)
...
 char *
 pcmk__time_format_hr(const char *format, const pcmk__time_hr_t *hr_dt)
 {
 #define DATE_LEN_MAX 128
-    const char *mark_s = NULL;
-    int scanned_pos = 0;
-    int printed_pos = 0;
-    int fmt_pos = 0;
+    int scanned_pos = 0; // How many characters of format have been parsed
+    int printed_pos = 0; // How many characters of format have been processed
```
[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/common/iso8601.c:1934]: I was looking into more changes but they aren't worth the hassle

File: lib/common/iso8601.c [hunk 6]
```diff
@@ -1971,10 +1988,8 @@ pcmk__time_format_hr(const char *format, const pcmk__time_hr_t *hr_dt)
                     fmt_pos = scanned_pos; // Show remaining string as-is
```
[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/common/iso8601.c:1990]: Ah what I really meant was passed as-is to strftime()

File: lib/common/iso8601.c [hunk 7]
```diff
@@ -2002,16 +2000,9 @@ pcmk__time_format_hr(const char *format, const pcmk__time_hr_t *hr_dt)
...
-#pragma GCC diagnostic push
-#pragma GCC diagnostic ignored "-Wformat-nonliteral"
-#endif
+        if (nano_digits != 0) {
             date_len += snprintf(&date_s[date_len], DATE_LEN_MAX - date_len,
-                                 nanofmt_s, nano_s);
-#ifdef HAVE_FORMAT_NONLITERAL
-#pragma GCC diagnostic pop
-#endif
+                                 "%.*s", nano_digits, nano_s);
```
[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/common/iso8601.c:2009]: This is for SNMP traps that require nanosecond formatting

File: lib/common/tests/iso8601/pcmk__time_format_hr_test.c (hunk 2 above)
Comment (kgaillot) [reply lib/common/tests/iso8601/pcmk__time_format_hr_test.c:80]: The documented syntax doesn't allow for it, so technically it's "undefined behavior". A literal percent is supposed to be doubled. The buffer overflow in our code was for this syntax though, so I wanted to keep a test for it.

File: lib/common/tests/iso8601/pcmk__time_format_hr_test.c (hunk 3 above)
Comment (kgaillot) [reply lib/common/tests/iso8601/pcmk__time_format_hr_test.c:70]: Yeah I figured another run would find something else

**Comment by kgaillot** [comment]
Updated comments and unit test per review

**Comment by kgaillot** [comment]
unit test fix

**Comment by kgaillot** [comment]
another unit test fix

**Comment by kgaillot** [comment]
hopefully last bsd unit test fix

**Review by nrwahl2 — APPROVED** [review]

**Pull request merged**